File: flux-basic-java/src/main/java/com/example/app/authentication/Role.java
package com.example.app.authentication;

import lombok.Getter;

@Getter
public enum Role {
    admin,
    owner(admin);

    private final Role[] assumedRoles;

    Role(Role... assumedRoles) {
        this.assumedRoles = assumedRoles;
    }

    public boolean matches(Role userRole) {
        if (userRole == null) {
            return false;
        }
        if (this == userRole) {
            return true;
        }
        for (Role assumedRole : userRole.getAssumedRoles()) {
            if (matches(assumedRole)) {
                return true;
            }
        }
        return false;
    }
}

File: flux-basic-kotlin/src/main/kotlin/com/example/app/authentication/Role.kt
package com.example.app.authentication

enum class Role(vararg val assumedRoles: Role) {
    MANAGER,
    ADMIN(MANAGER),
    OWNER(ADMIN);

    fun matches(userRole: Role?): Boolean {
        if (userRole == null) {
            return false
        }
        if (this == userRole) {
            return true
        }
        for (assumedRole in userRole.assumedRoles) {
            if (matches(assumedRole)) {
                return true
            }
        }
        return false
    }
}
